### ISSUE TITLE
Avro java serialized deserialization not working

### ISSUE THREAD

**Issue by dhiaayachi** [actgardner/gogen-avro#130]
I'm working on a project that need to have a java service and intended to use avro for service to service communication as follow:

                      JAVA APP----avro Req (byte serialized)----> GO APP
                      <-----avro Rsp (byte serialized)----

I had trouble deserializing Java serialized avro and vice versa deserializing go serialized response. After investigation what I see is the following:

- Java avro serializer when serializing an avro message add a fixed header (2 bytes) (0xC301) to annouce the avro version at the beginning of the binary message
- Java avro serializer add a long (8 bytes) (ex: 0C947F601DE7CE84)  which is a magic number to announce the schema version ( it's calculated based on the schema string)

If I remove those both headers the message is working correctly and the serialization/deserialization is done correctly on both sides.

I'm not an avro expert but this seem to be part of the avro specification and I'm wondering if this is something that would be added and that you will agree to merge. If so, I can work on it with a bit of help.

I would like also to mention that it also not supported by goavro

**Comment by dhiaayachi** [comment]
Update I see that the crc64 is implemented commit aac6fb64e but I don't see how to use it, specially using the command line tool, am I missing something?

**Comment by dhiaayachi** [comment]
I was able to put the code together crc + canonical form to calculate the the same fingerprint for my schema in java and go side. The idea I have is to change the code generator to include the fingerprint in the generated struct and modify the Serialize/Deserialize func to add/check the fingerprint.
I will need help to understand how the code generation work.

**Comment by actgardner** [comment]
You're describing Avro's Single Object Encoding (https://avro.apache.org/docs/current/spec.html#single_object_encoding) which is part of the spec we're working on support for (see #111)

> The idea I have is to change the code generator to include the fingerprint in the generated struct and modify the Serialize/Deserialize func to add/check the fingerprint.

I think putting the fingerprint in the generated code is a good idea. Rather than modifying the Serialize/Deserialize methods we'll probably add a new Reader/Writer pair that handle framing the objects. I can take a swing at it this evening.

**Comment by dhiaayachi** [comment]
Yes that exactly it.
I already did a patch that use the canonical package (with some slight modification to it) and modified the record.go template to add the headers. The issue I'm having is that I'm not able to configure my environment properly to work with my fork So I can fix the failing tests. See commit below with the modification I did.
https://github.com/dhiaayachi/gogen-avro/commit/bb8d8afb0e3c9937a262b83cb149253ebe9e3190
It use the Serialize/Deserialize method for now but I can update it to use a different pair of reader/writer
